git/ls_files.go

```diff
@@ -0,0 +1,90 @@
+package git
+
+import (
+	"bufio"
+	"strings"
+	"path"
+	"io/ioutil"
+
+	"github.com/git-lfs/git-lfs/errors"
+	"github.com/git-lfs/git-lfs/tools"
+	"github.com/rubyist/tracerx"
+)
+
+type lsFileInfo struct {
+	BaseName string
+	FullPath string
+}
+type LsFiles struct {
+	Files map[string]*lsFileInfo
+	FilesByName map[string][]*lsFileInfo
+}
+
+func NewLsFiles(workingDir string, standardExclude bool) (*LsFiles, error) {
+
+	args := []string{
+		"ls-files",
+		"-z", // Use a NUL separator. This also disables the escaping of special characters.
+		"--others",
+		"--cached",
+	}
+
+	if standardExclude {
+		args = append(args, "--exclude-standard")
+	}
+	cmd := gitNoLFS(args...)
+	cmd.Dir = workingDir
+
+    tracerx.Printf("NewLsFiles: running in %s git %s",
+			workingDir, strings.Join(args, " "))
+
+	// Capture stdout and stderr
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, err
+	}
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return nil, err
+	}
+
+	scanner := bufio.NewScanner(stdout)
+	scanner.Split(tools.SplitOnNul)
+
+	if err := cmd.Start(); err != nil {
+		return nil, err
+	}
+
+	rv := &LsFiles {
+		Files: make(map[string]*lsFileInfo),
+		FilesByName: make(map[string][]*lsFileInfo),
+	}
+
+	// Setup a goroutine to drain stderr as large amounts of error output may cause
+	// the subprocess to block.
+	errorMessages := make(chan []byte)
+	go func() {
+		msg, _ := ioutil.ReadAll(stderr)
+		errorMessages <- msg
+	}()
+
+	// Read all files
+	for scanner.Scan() {
+		base := path.Base(scanner.Text())
+		finfo := &lsFileInfo {
+			BaseName: base,
+			FullPath: scanner.Text(),
+		}
+		rv.Files[scanner.Text()] = finfo
+		rv.FilesByName[base] = append(rv.FilesByName[base], finfo)
```

Comment:
I think this approach has the real possibility of using a lot of memory on large repositories, since we're storing data twice for each file instead of operating incrementally.  I think we'd want to continue to do things incrementally like we do for the current walker.

Reply:
Yeah, I considered that when I did it. For my repo (which IMO is quite large w/ a working tree of ~300k files), `ls-files -z -o --cached` produced around 20 MiB of data. I made a completely untested tradeoff for speed here, with the assumption that a callback or channel based approach would be slower. Perhaps the most prudent thing to do would be for me to try it with an iterative interface, and see what happens to the execution time/memory consumption.

Reply:
Largest git installation I worked on was 1M files with 1.5TB `.git/objects` (before git-lfs existed) and this was much beyond what any sane person would do. So if what you did works reasonable well for 300K files, I believe that's Good Enough to leave as is.

Reply:
Okay, let's go with it.  It's going to scale better than the current status quo, and it's not too difficult to fix if it doesn't.